### DISCUSSION TITLE
Dimensionality reduction algorithm (e.g. t-SNE*)

### DISCUSSION THREAD

**Discussion by chuanqisun** [cozodb/cozo#83] [Ideas]
In the spirit of being the "hippocampus for AI", Cozo already supports vectors as a first class data type, along with HNSW index for high performance vector search. Since high dimensional vectors are central in modern ML algorithms, a high performance [dimensionality reduction algorithm](https://en.wikipedia.org/wiki/Dimensionality_reduction) would be a very attractive addition to the current [algorithm toolbox](https://docs.cozodb.org/en/latest/algorithms.html).

From an end user's perspective, such algorithm could project high dimensional vectors into a 2D/3D space, which can be fed into plotting libraries for visual cluster analysis and story telling.

Additionally, since Cozo is available via WASM, it is possible to build a web-based demo that shows how fast Cozo can perform such algorithms and hook up the output to some web-based visualization library (such as D3.js). This would provide an intuitive and compelling story on the flexibility of Cozo in working with vectors, which in my opinion is a unique reason why people would prefer Cozo to other solutions.

*t-SNE is just an example. There are probably many options out there as solution candidates.

**Comment by Madd0g** [comment]
+1

I learned about [UMAP](https://arxiv.org/abs/1802.03426) and immediately came to see if anyone asked this.

I found a rust implementation of the algo - https://crates.io/crates/annembed
